src/libsystemd-network/sd-lldp-rx.c

```diff
@@ -491,7 +492,50 @@ int sd_lldp_rx_get_neighbors(sd_lldp_rx *lldp_rx, sd_lldp_neighbor ***ret) {
         return k;
 }
 
-int sd_lldp_rx_set_neighbors_max(sd_lldp_rx *lldp_rx, uint64_t m) {
+_public_ int sd_lldp_rx_get_neighbors_json(sd_lldp_rx *lldp, JsonVariant ***ret) {
+        int i, j, r, n = 0;
+        _cleanup_free_ sd_lldp_neighbor **l = NULL;
+        JsonVariant **v = NULL;
+
+        assert_return(lldp, -EINVAL);
+        assert_return(ret, -EINVAL);
+
+        r = sd_lldp_rx_get_neighbors(lldp, &l);
+        if (r < 0)
+                return r;
+
+        n = r;
+
+        v = new0(JsonVariant*, n);
+        if (!v)
+                return -ENOMEM;
+
+        for (i = 0; i < n; i++) {
+                r = sd_lldp_neighbor_build_json(l[i], v + i);
+                if (r < 0)
+                        goto clear;
+                else
+                        sd_lldp_neighbor_unref(l[i]);
+        }
+
+        *ret = v;
+
+        for (i = 0; i < n; j++)
+
+        return n;
+
+clear:
+        for (j = i; j < n; j++)
+                sd_lldp_neighbor_unref(l[j]);
+
+        for (j = 0; j < i; j++)
+                json_variant_unrefp(v + j);
```

Comment:
use CLENAUP_ARRAY() with json_variant_unref_many

Reply:
While `CLEANUP_ARRAY` is indeed a good tool, the initialized part of this array needs to be cleaned only if something goes wrong (see `goto clear`). I can't figure out how make `CLEANUP_ARRAY` work here. :( Any ideas?